File: services/server/src/main/java/com/yaytsa/server/infrastructure/persistence/repository/AlbumRepository.java
package com.yaytsa.server.infrastructure.persistence.repository;

import com.yaytsa.server.infrastructure.persistence.entity.AlbumEntity;
import java.util.Collection;
import java.util.List;
import java.util.UUID;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;
import org.springframework.stereotype.Repository;

@Repository
public interface AlbumRepository extends JpaRepository<AlbumEntity, UUID> {
  List<AlbumEntity> findByArtistId(UUID artistId);

  @Query("SELECT a FROM AlbumEntity a WHERE a.artist.id = :artistId ORDER BY a.item.sortName ASC")
  List<AlbumEntity> findByAlbumArtistIdOrderBySortNameAsc(@Param("artistId") UUID artistId);

  @Query(
      "SELECT a FROM AlbumEntity a LEFT JOIN FETCH a.artist LEFT JOIN FETCH a.item WHERE a.itemId ="
          + " :itemId")
  AlbumEntity findByIdWithArtist(@Param("itemId") UUID itemId);

  @Query("SELECT COUNT(a) FROM AlbumEntity a WHERE a.artist.id = :artistId")
  long countByArtistId(@Param("artistId") UUID artistId);

  @Query(
      "SELECT a FROM AlbumEntity a "
          + "LEFT JOIN FETCH a.artist "
          + "LEFT JOIN FETCH a.item "
          + "WHERE a.itemId IN :itemIds")
  List<AlbumEntity> findAllByIdInWithArtist(@Param("itemIds") Collection<UUID> itemIds);
}

File: yay-tsa-v2/infra-persistence/library/src/main/kotlin/dev/yaytsa/persistence/library/repository/AlbumRepository.kt
package dev.yaytsa.persistence.library.repository

import dev.yaytsa.persistence.library.entity.AlbumJpa
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.query.Param
import java.util.UUID

interface AlbumRepository : JpaRepository<AlbumJpa, UUID> {
    fun findByArtistId(artistId: UUID): List<AlbumJpa>

    @Query(
        value =
            "SELECT * FROM core_v2_library.albums " +
                "WHERE metadata_checked_at IS NULL ORDER BY entity_id LIMIT :limit OFFSET :offset",
        nativeQuery = true,
    )
    fun findByMetadataCheckedAtIsNull(
        limit: Int,
        offset: Int,
    ): List<AlbumJpa>

    // Image-driven artwork backfill: albums with no primary image row, regardless of whether
    // metadata_checked_at is already stamped. Self-healing — when a cover source improves, art-less
    // albums are reconsidered on the next cycle without a manual metadata_checked_at reset.
    @Query(
        value =
            "SELECT a.* FROM core_v2_library.albums a " +
                "WHERE NOT EXISTS (" +
                "SELECT 1 FROM core_v2_library.images i WHERE i.entity_id = a.entity_id AND i.is_primary) " +
                "ORDER BY a.entity_id LIMIT :limit OFFSET :offset",
        nativeQuery = true,
    )
    fun findWithoutPrimaryImage(
        limit: Int,
        offset: Int,
    ): List<AlbumJpa>

    fun findByReleaseDateBetween(
        from: java.time.LocalDate,
        to: java.time.LocalDate,
        pageable: org.springframework.data.domain.Pageable,
    ): List<AlbumJpa>

    @Query(
        "SELECT a.artistId AS artistId, COUNT(a) AS albumCount " +
            "FROM AlbumJpa a WHERE a.artistId IN :artistIds GROUP BY a.artistId",
    )
    fun countAlbumsByArtistIds(
        @Param("artistIds") artistIds: Collection<UUID>,
    ): List<AlbumCountByArtist>
}

interface AlbumCountByArtist {
    fun getArtistId(): UUID

    fun getAlbumCount(): Long
}
